sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/client/SentinelApiClient.java

```diff
@@ -121,19 +136,139 @@ protected boolean isRedirectable(final String method) {
+                logger.warn("httpPostContent encode entity error: {}", params, e);
+                return null;
+            }
+        }
+        return httpPost;
+    }
+    
+    private String urlEncode(String str) {
+        try {
+            return URLEncoder.encode(str, DEFAULT_CHARSET.name());
+        } catch (UnsupportedEncodingException e) {
+            logger.info("encode string error: {}", str, e);
+            return null;
+        }
+    }
+    
+    private String getBody(HttpResponse response) throws Exception {
+        Charset charset = null;
+        try {
+            String contentTypeStr = response.getFirstHeader("Content-type").getValue();
+            if (StringUtil.isNotEmpty(contentTypeStr)) {
+                ContentType contentType = ContentType.parse(contentTypeStr);
+                charset = contentType.getCharset();
+            }
+        } catch (Exception ignore) {
+        }
+        return EntityUtils.toString(response.getEntity(), charset != null ? charset : DEFAULT_CHARSET);
+    }
+    
+    /**
+     * With no param
+     * 
+     * @param ip
+     * @param port
+     * @param api
+     * @return
+     */
+    private CompletableFuture<String> executeCommand(String ip, int port, String api) {
+        return executeCommand(ip, port, api, null);
+    }
+    
+    /**
+     * No app specified, force to GET
+     * 
+     * @param ip
+     * @param port
+     * @param api
+     * @param params
+     * @return
+     */
+    private CompletableFuture<String> executeCommand(String ip, int port, String api, Map<String, String> params) {
+        return executeCommand(null, ip, port, api, params);
+    }
 
-    private CompletableFuture<String> executeCommand(String command, URI uri) {
+    /**
+     * Prefer to execute request using POST
+     * 
+     * @param app
+     * @param ip
+     * @param port
+     * @param api
+     * @param params
+     * @return
+     */
+    private CompletableFuture<String> executeCommand(String app, String ip, int port, String api, Map<String, String> params) {
         CompletableFuture<String> future = new CompletableFuture<>();
-        if (StringUtil.isBlank(command) || uri == null) {
+        if (StringUtil.isBlank(ip) || StringUtil.isBlank(api)) {
             future.completeExceptionally(new IllegalArgumentException("Bad URL or command name"));
             return future;
         }
-        final HttpGet httpGet = new HttpGet(uri);
-        httpClient.execute(httpGet, new FutureCallback<HttpResponse>() {
+        StringBuilder urlBuilder = new StringBuilder();
+        urlBuilder.append("http://");
+        urlBuilder.append(ip).append(':').append(port).append('/').append(api);
+        if (params == null || params.size() == 0) {
+            // No parameter, using GET
+            return executeCommand(new HttpGet(urlBuilder.toString()));
+        }
+        boolean supportPost = StringUtil.isNotEmpty(app) && Optional.ofNullable(appManagement.getDetailApp(app))
```

Comment:
Can we add a flag in the parameter of `executeCommand` indicating whether the API is POST-based? Using null `app` might be confusing.

Reply:
Append a new parameter to indicate posting intent